libnetwork/osl/interface_linux.go

```diff
@@ -276,22 +323,232 @@ func (n *Namespace) AddInterface(ctx context.Context, srcName, dstPrefix string,
 	if err != nil {
 		return fmt.Errorf("failed to set link up: %v", err)
 	}
+	log.G(ctx).Debug("link has been set to up")
 
 	// Set the routes on the interface. This can only be done when the interface is up.
 	if err := setInterfaceRoutes(ctx, nlh, iface, i); err != nil {
 		return fmt.Errorf("error setting interface %q routes to %q: %v", iface.Attrs().Name, i.Routes(), err)
 	}
 
+	// Wait for the interface to be up and running (or a timeout).
+	up, err := waitForIfUpped(ctx, newNs, iface.Attrs().Index)
+	if err != nil {
+		return err
+	}
+
+	// If the interface is up, send unsolicited ARP/NA messages if necessary.
+	if up {
+		if err := n.advertiseAddrs(ctx, iface.Attrs().Index, i, nlh); err != nil {
+			return fmt.Errorf("failed to advertise addresses: %w", err)
+		}
+	}
+
 	n.mu.Lock()
 	n.iFaces = append(n.iFaces, i)
 	n.mu.Unlock()
 
 	return nil
 }
 
+func waitForIfUpped(ctx context.Context, ns netns.NsHandle, ifIndex int) (bool, error) {
+	update := make(chan netlink.LinkUpdate, 100)
+	upped := make(chan struct{})
+	opts := netlink.LinkSubscribeOptions{
+		ListExisting: true, // in case the link is already up
+		ErrorCallback: func(err error) {
+			select {
+			case <-upped:
+				// Ignore errors sent after the upped channel is closed, the netlink
+				// package sends an EAGAIN after it closes its netlink socket when it
+				// sees this channel is closed. (No message is ever sent on upped.)
+				return
+			default:
+			}
+			log.G(ctx).WithFields(log.Fields{
+				"ifi":   ifIndex,
+				"error": err,
+			}).Info("netlink error while waiting for interface up")
+		},
+	}
+	if ns.IsOpen() {
+		opts.Namespace = &ns
+	}
+	if err := nlwrap.LinkSubscribeWithOptions(update, upped, opts); err != nil {
+		return false, fmt.Errorf("failed to subscribe to link updates: %w", err)
+	}
+
+	timer := time.NewTimer(5 * time.Second)
+	defer timer.Stop()
+	defer close(upped)
```

Comment:
Is there a chance that 
https://github.com/moby/moby/blob/f95e4b3d22260eb3322b4be13d391b8db2d4199f/vendor/github.com/vishvananda/netlink/link_linux.go#L2452

won't return immediately after `s.Close()`?

The goroutine inside `linkSubscribeAt` closes the `update` channel before it exists. 
Wondering if we should wait for the `update` after requesting the `linkSubscribeAt` to end, so that we're sure that the goroutine actually exited?

```suggestion
	defer func() {
		close(upped)
		select {
		  case <-update:
		  case time.After(time.Second * 3):
		     // log warning/error
		}
	}
```

Reply:
I've done that ... slightly modified because I think another message on the `update` could race the close - so the select would still terminate before the netlink goroutine terminates and closes the channel.